packages/contracts-bedrock/contracts/echidna/FuzzResourceMetering.sol

```diff
@@ -0,0 +1,133 @@
+        // Part 1: we cache the current param values and do some basic checks on them.
+        uint256 cachedPrevBaseFee = uint256(params.prevBaseFee);
+        uint256 cachedPrevBoughtGas = uint256(params.prevBoughtGas);
+        uint256 cachedPrevBlockNum = uint256(params.prevBlockNum);
+
+        // check that the last block's base fee hasn't dropped below the minimum
+        if (cachedPrevBaseFee < uint256(MINIMUM_BASE_FEE)) {
+            failedNeverBelowMinBasefee = true;
+        }
+        // check that the last block didn't consume more than the max amount of gas
+        if (cachedPrevBoughtGas > uint256(MAX_RESOURCE_LIMIT)) {
+            failedMaxGasPerBlock = true;
+        }
+
+        // Part2: we perform the gas burn
+
+        // force the gasToBurn into the correct range based on whether we intend to
+        // raise or lower the basefee after this block, respectively
+        uint256 gasToBurn;
+        if (_raiseBaseFee) {
+            gasToBurn = bound(
+                _gasToBurn,
+                uint256(TARGET_RESOURCE_LIMIT),
+                uint256(MAX_RESOURCE_LIMIT)
+            );
+        } else {
+            gasToBurn = bound(_gasToBurn, 0, uint256(TARGET_RESOURCE_LIMIT));
+        }
+
+        _burnInternal(uint64(gasToBurn));
+
+        // Part 3: we run checks and modify our invariant flags based on the updated params values
+
+        // Calculate the maximum allowed baseFee change (per block)
+        uint256 maxBasefeeChange = cachedPrevBaseFee / uint256(BASE_FEE_MAX_CHANGE_DENOMINATOR);
+
+        // If the last block used more than the target amount of gas (and there were no
+        // empty blocks in between), ensure this block's basefee increased, but not by
+        // more than the max amount per block
+        if (
+            (cachedPrevBoughtGas > uint256(TARGET_RESOURCE_LIMIT)) &&
+            (uint256(params.prevBlockNum) - cachedPrevBlockNum == 1)
+        ) {
+            failedRaiseBasefee = failedRaiseBasefee || (params.prevBaseFee <= cachedPrevBaseFee);
+            failedMaxRaiseBasefeePerBlock =
+                failedMaxRaiseBasefeePerBlock ||
+                ((uint256(params.prevBaseFee) - cachedPrevBaseFee) < maxBasefeeChange);
+        }
+
+        // If the last blocked used less than the target amount of gas ensure this block's basefee
+        // decreased, but not by more than the max amount
+        if (
+            (cachedPrevBoughtGas < uint256(TARGET_RESOURCE_LIMIT)) ||
+            (uint256(params.prevBlockNum) - cachedPrevBlockNum > 1)
+        ) {
+            failedLowerBasefee =
+                failedLowerBasefee ||
+                (uint256(params.prevBaseFee) > cachedPrevBaseFee);
+            if (params.prevBlockNum - cachedPrevBlockNum == 1) {
+                failedMaxLowerBasefeePerBlock =
+                    failedMaxLowerBasefeePerBlock ||
+                    ((cachedPrevBaseFee - uint256(params.prevBaseFee)) < maxBasefeeChange);
+            }
```

Comment:
Could make the code below into an `else` block

Reply:
I considered that, but I didn't feel good about it, because AFAICT, it could also be possible that `params.prevBlockNum - cachedPrevBlockNum === 0` here.

Or am I missing something?